### PR TITLE
infra: support reloadable EnvoyGateway configuration

### PR DESCRIPTION
**Pull request by zirain**
fixes: https://github.com/envoyproxy/gateway/issues/3209

### CODE REVIEW

File: internal/envoygateway/config/runner/runner.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	return runnerName
+}
+
+func (r *Runner) Start(ctx context.Context) error {
+	if r.cfgPath == "" {
+		r.logger.Info("no config file provided, skipping config watcher")
+		return nil
+	}
+
+	watcher, err := fsnotify.NewWatcher()
```
Comment (shawnh2) [inline internal/envoygateway/config/runner/runner.go:43]: Is possible to reuse the notifier under `internal/provider/file/notifier.go` ?

Comment (zirain) [reply internal/envoygateway/config/runner/runner.go:43]: still WIP, will try it later.

Comment (zirain) [reply internal/envoygateway/config/runner/runner.go]: IMO `internal/provider/file/notifier.go` had not user-firendly API, I will try to use lib from hugo in this PR, cause it's not the key point.

**Comment by zirain** [comment]
failed because [ExtensionAPISettings](https://github.com/envoyproxy/gateway/blob/f5693f6bc1bba4d770e0fea9d9e2b64617f2140b/api/v1alpha1/envoygateway_types.go#L165) didn't updated. It's able to add new watcher after `EnvoyGateway` updated, but cannot remove watcher after `EnvoyGateway` updated.

File: internal/common/runner.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package common
```
Comment (arkodg) [inline internal/common/runner.go:6]: prefer `runner` package under `internal/runner`

File: internal/envoygateway/config/loader/Loader.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		defer w.Close()
+		for {
+			select {
+			case events, ok := <-w.Events:
+				if !ok {
+					return
+				}
+
+				for _, e := range events {
+					r.logger.Info("receive fsnotify events", "name", e.Name, "op", e.Op.String())
```
Comment (arkodg) [inline internal/envoygateway/config/loader/Loader.go:71]:
```suggestion
					r.logger.Info("received fsnotify events", "name", e.Name, "op", e.Op.String())
```

File: internal/envoygateway/config/loader/Loader.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+		r.logger.Info("no config file provided, skipping config watcher")
+		return nil
+	}
+
+	w, err := watcher.New(500*time.Millisecond, 500*time.Millisecond, false)
+	if err != nil {
+		r.logger.Error(err, "failed to create fsnotify watcher")
+		return err
+	}
+	r.logger.Info("watch file changes", "path", r.cfgPath)
```
Comment (arkodg) [inline internal/envoygateway/config/loader/Loader.go:54]:
```suggestion
	r.logger.Info("watching for changes to the EnvoyGateway configuration", "path", r.cfgPath)
```

File: internal/envoygateway/config/loader/Loader.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+				}
+
+				for _, e := range events {
+					r.logger.Info("receive fsnotify events", "name", e.Name, "op", e.Op.String())
+				}
+
+				// Load the config file.
+				eg, err := config.Decode(r.cfgPath)
+				if err != nil {
+					r.logger.Error(err, "failed to decode config file", "name", r.cfgPath)
```
Comment (arkodg) [inline internal/envoygateway/config/loader/Loader.go:77]: should we silently fail this ? imo we should follow the same flow we do during startup

File: internal/gatewayapi/runner/runner.go [hunk 6]
```diff
@@ -43,17 +46,21 @@ const (
 	serveTLSCaFilename   = "/certs/ca.crt"
 )
 
+var _ common.Runner = &Runner{}
+
 type Config struct {
-	config.Server
+	ServerCfg         *config.Server
```
Comment (arkodg) [inline internal/gatewayapi/runner/runner.go:52]: curious why :) ?

File: internal/gatewayapi/runner/runner.go [hunk 7]
```diff
@@ -120,7 +138,7 @@ func (r *Runner) subscribeAndTranslate(ctx context.Context) {
 			r.Logger.Info("received an update")
 			val := update.Value
 			// There is only 1 key which is the controller name
-			// so when a delete is triggered, delete all IR keys
+			// so when delete is triggered, delete all IR keys
```
Comment (arkodg) [inline internal/gatewayapi/runner/runner.go:141]: the prev grammar was correct, lets undo this

File: internal/gatewayapi/runner/runner.go (hunk 6 above)
Comment (zirain) [reply internal/gatewayapi/runner/runner.go:52]: Using pointers instead of value copies?

File: internal/provider/kubernetes/kubernetes.go [hunk 8]
```diff
@@ -110,7 +111,7 @@ func New(cfg *rest.Config, svr *ec.Server, resources *message.ProviderResources)
 	// Emit elected & continue with deployment of infra resources
 	go func() {
 		<-mgr.Elected()
-		close(svr.Elected)
```
Comment (arkodg) [inline internal/provider/kubernetes/kubernetes.go:113]: logic is changing here ...

File: internal/infrastructure/kubernetes/ratelimit_infra.go [hunk 9]
```diff
@@ -52,10 +52,6 @@ func (i *Infra) CreateOrUpdateRateLimitInfra(ctx context.Context) error {
 
 // DeleteRateLimitInfra removes the managed kube infra, if it doesn't exist.
 func (i *Infra) DeleteRateLimitInfra(ctx context.Context) error {
-	if err := ratelimit.Validate(ctx, i.Client.Client, i.EnvoyGateway, i.Namespace); err != nil {
```
Comment (arkodg) [inline internal/infrastructure/kubernetes/ratelimit_infra.go:55]: curious why this was needed

File: internal/provider/kubernetes/kubernetes.go (hunk 8 above)
Comment (zirain) [reply internal/provider/kubernetes/kubernetes.go:113]: a channel can only be closed once

Comment (arkodg) [reply internal/provider/kubernetes/kubernetes.go:113]: sg, do we have a leader election test ?
cc @alexwo

Comment (zirain) [reply internal/provider/kubernetes/kubernetes.go:113]: tracked in https://github.com/envoyproxy/gateway/issues/4420

File: internal/infrastructure/kubernetes/ratelimit_infra.go (hunk 9 above)
Comment (zirain) [reply internal/infrastructure/kubernetes/ratelimit_infra.go:55]: will be fixed in https://github.com/envoyproxy/gateway/pull/4430

File: internal/xds/server/runner/runner.go [hunk 10]
```diff
@@ -82,29 +84,37 @@ func (r *Runner) Start(ctx context.Context) (err error) {
...
-	r.Logger.Info("started")
+	r.logger.Info("started")
 	return
 }
 
+func (r *Runner) Reload(serverCfg *config.Server) error {
+	r.logger = serverCfg.Logger.WithName("xds-server-runner")
+	r.logger.Info("reloaded")
+
+	// TODO: Implement reload logic
```
Comment (zhaohuabing) [inline internal/xds/server/runner/runner.go:103]: Hi @zirain Is this PR still a draft?

Comment (zirain) [reply internal/xds/server/runner/runner.go:103]: nit, should be removed.

**Comment by zhaohuabing** [comment]
Dealing with configuration reload for each runner can be cumbersome, especially as the `EnvoyGateway` may introduce additional settings in the future. Would it be more straightforward to restart all runners instead of reloading the configuration individually?

**Comment by zirain** [comment]
close this first in favor of https://github.com/envoyproxy/gateway/pull/4451, maybe came back later.

**Pull request closed without merging**